test/ajax.js

```diff
@@ -19,12 +19,104 @@ QUnit.test( "jQuery.ajax() deprecations on jqXHR", function( assert ) {
 			} )
 			.complete( function() {
 				assert.ok( true, "ajax complete" );
+			} )
+			.catch( jQuery.noop );
+	} ).then( function() {
+		done();
+	} );
+
+} );
+
+[ " - Same Domain", " - Cross Domain" ].forEach( function( label, crossDomain ) {
+
+	// The JSON-to-JSONP auto-promotion behavior is gone in jQuery 4.0 and as
+	// it has security implications, we don't want to restore the legacy behavior.
+	QUnit[ jQueryVersionSince( "4.0.0" ) ? "skip" : "test" ](
+			"jQuery.ajax() JSON-to-JSONP auto-promotion" + label, function( assert ) {
+
+		assert.expect( 5 );
+
+		var done = assert.async(),
+			tests = [
+				function() {
+					return expectNoWarning( assert, "dataType: \"json\"",
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json" ),
+								crossDomain: crossDomain,
+								dataType: "json"
+							} ).catch( jQuery.noop );
+						}
+					);
+				},
 
-				// Wait for expectWarning to complete
-				setTimeout( done, 1 );
+				function() {
+					return expectWarning( assert, "dataType: \"json\", URL callback", 1,
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json?callback=?" ),
+								crossDomain: crossDomain,
+								dataType: "json"
+							} ).catch( jQuery.noop );
+						}
+					);
+				},
+
+				function() {
+					return expectWarning( assert, "dataType: \"json\", data callback", 1,
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json" ),
+								crossDomain: crossDomain,
+								data: "callback=?",
+								dataType: "json"
+							} ).catch( jQuery.noop );
+						}
+					);
+				},
+
+				function() {
+					return expectNoWarning( assert, "dataType: \"jsonp\", URL callback",
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json?callback=?" ),
+								crossDomain: crossDomain,
+								dataType: "jsonp"
+							} ).catch( jQuery.noop );
+						}
+					);
+				},
+
+				function() {
+					return expectNoWarning( assert, "dataType: \"jsonp\", data callback",
+						function() {
+							return jQuery.ajax( {
+								url: url( "data/null.json" ),
+								crossDomain: crossDomain,
+								data: "callback=?",
+								dataType: "jsonp"
+							} ).catch( jQuery.noop );
+						}
+					);
+				}
+			];
+
+		// Invoke tests sequentially as they're async and early tests could get warnings
+		// from later ones.
+		function run( tests ) {
+			var test = tests[ 0 ];
+			return test().then( function() {
+				if ( tests.length > 1 ) {
+					return run( tests.slice( 1 ) );
+				}
 			} );
-	} );
+		}
 
+		run( tests )
+			.then( function() {
+				done();
+			} );
+	} );
```

Comment:
This is quite a lot of boilerplate but we don't have that many tests so I'm not sure if this is the moment to generalize it... If there's a way to simplify, I'm all ears. :)

Reply:
I'm fine with it, if it really gets out of hand we can refactor some other time.